graph/graph.go

```diff
@@ -3,12 +3,97 @@ package graph
 import (
 	"fmt"
 	"strings"
+
+	proto "github.com/golang/protobuf/proto"
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/protobuf/testing/protocmp"
 )
 
-// Graph is a map-basd adjacency list for storing a lineage graph
+// Graph is a map-basd adjacency list for storing a lineage graph.
+// It contains the main adjacency list, Edges, for normal edges.
+// For projector argument edges, it contains the ArgumentEdges adjacency list.
+// It also contains a lookup dictionary of all nodes in the graph.
 type Graph struct {
-	Graph map[IsID][]IsID
-	Nodes map[IsID]Node
+	Edges         map[IsID][]IsID
+	ArgumentEdges map[IsID][]IsID
+	Nodes         map[IsID]Node
+}
+
+// addNode adds a node to the appropriate graph. All nodes are added to the Nodes list.
+// ProjectorDef nodes are added to both Edges and ArgumentEdges adjacency lists.
+// Arguments of a projector are added as descendants to their projector only in the ArgumentEdges adjacency list.
+// Sub-mappings of a projector are added as descendants to their projector only in the Edges adjacency list.
+func (g Graph) addNode(node Node, descendant Node, isArg bool, isNew bool) error {
```

Comment:
Is there a reason that we are using the Graph value as the receiver?

Reply:
Also, we might need some tests for this function.

Reply:
I had considered adding a node as a behavior "belonging" to the graph, but on review think it works better if associated with the rest of the graph generation functions. I've removed Graph as the receiver.
There are tests for this function in generate_test.go; I had placed them in the wrong file. I will keep them there since I've now moved addNode to generate.go anyways.